REST high-level client: add Cluster Health API

Van0SS opened the pull request:
Relates to #27205

It's not a final PR, there are things to discuss and tasks to finish. Want to be sure that I am on a right track.

Questions:
- There is formatting output parameter 'level'. Currently, it is used only in REST API. Do we need it in High level client? **UPD: added**
  - It can allow to decrease traffic and return empty 'indices' map if the level is set to 'cluster'. The same for shards.

- Found out that API parameter 'master_timeout' isn't [documented](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-health.html) but present in ClusterHealthRequest and parsed in REST Controller. Is it as it should be? **UPD: added**

For discussion in this PR:
- Serialize in XContent object name (in this case 'index' and 'shardId') **UPD: removed**
  - Pros: Objects can be XContented without level-up name and still have all fields; Easier to parse (including for any other clients)
  - Cons: Non-breaking but API change; A little overhead

Suggestions to implement:
- Set default 'level' to 'cluster' in REST ~~client~~ CONTROLLER level, not like now - in toXContent **UPD: better do it on transport level but not now**
  - Pros: Default XContent represent the full state of POJO
  - Cons: Finding a nice way to do it

Thanks in advance for reviewing, I understand that there are some changes in the existing code.

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
Pinging @elastic/es-core-infra

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/RequestTests.java:
```diff
@@ -1325,6 +1328,25 @@ public void testClusterPutSettings() throws IOException {
+        setRandomTimeout(healthRequest::timeout, AcknowledgedRequest.DEFAULT_ACK_TIMEOUT, expectedParams);
+        setRandomMasterTimeout(healthRequest, expectedParams);
+        expectedParams.put("level", "shards");
+        // Default value in ClusterHealthRequest is NONE but in Request.Params::withWaitForActiveShards is DEFAULT
```
I see, you may have to make changes here like we already did in the 6.x branch where we have a different default for some API.

Van0SS replied:
Can you please clarify it? Should I set in ClusterHealthRequest field `waitForActiveShards = ActiveShardCount.DEFAULT`?

javanna replied:
I mean to adapt the test, doing something along these lines (from the 6.x branch): https://github.com/elastic/elasticsearch/blob/6.x/client/rest-high-level/src/test/java/org/elasticsearch/client/RequestTests.java#L1649 .

javanna commented on server/src/main/java/org/elasticsearch/common/util/CollectionUtils.java:
```diff
@@ -427,4 +444,41 @@ public static int sortAndDedup(final BytesRefArray bytes, final int[] indices) {
+            return (int) ((float) expectedSize / 0.75F + 1.0F);
+        }
+        return Integer.MAX_VALUE; // any large value
+    }
```
I would refrain from adding these to be honest, especially if only used in our client/parsing code.

nik9000 replied:
++

For the most part we're ok with `new HashMap<>(expectedSize);`

nik9000 commented on server/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java:
```diff
@@ -263,4 +360,29 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+            numberOfInFlightFetch == that.numberOfInFlightFetch &&
+            delayedUnassignedShards == that.delayedUnassignedShards &&
+            timedOut == that.timedOut &&
+            Objects.equals(clusterName, that.clusterName) &&
```
Can you put these in the same order as `hashCode`?

javanna commented:
@Van0SS also, answers to your questions:

> There is formatting output parameter 'level'. Currently, it is used only in REST API. Do we need it in High level client?

Yes we do. If we don't allow the high-level client to provide this parameter, it will only ever be possible to retrieve cluster level view (default) using the high-level REST client. We need to add it as a field to `ClusterHealthRequest` and document that it's only used by the high-level REST client. We could also think of moving the filtering logic to the transport action rather than the REST layer, given that we retrieve everything at the moment but only print out at the REST layer what was requested. But I would not do it now. We will just need to be carefully to provide the proper params to the response#toXContent when testing. Let me know if you have questions on that.

> Found out that API parameter 'master_timeout' isn't documented but present in ClusterHealthRequest and parsed in REST Controller. Is it as it should be?

We probably forgot, could you also add that parameter to the SPEC? You can do it either as part of this PR or as a separate PR that could go in first if you prefer.

nik9000 commented on server/src/main/java/org/elasticsearch/cluster/health/ClusterIndexHealth.java:
```diff
@@ -197,22 +274,56 @@ public XContentBuilder toXContent(final XContentBuilder builder, final Params pa
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        ClusterIndexHealth that = (ClusterIndexHealth) o;
+        return numberOfShards == that.numberOfShards &&
```
Can you put these in the same order as hashcode?

nik9000 commented on server/src/main/java/org/elasticsearch/cluster/health/ClusterIndexHealth.java:
```diff
@@ -22,19 +22,86 @@
+            });
+
+    public static final ObjectParser.NamedObjectParser<ClusterShardHealth, Void> SHARD_PARSER =
+        (XContentParser p, Void c, String nameIgnored) -> ClusterShardHealth.fromXContent(p);
```
Thanks for calling it `nameIgnored`.

nik9000 commented on server/src/main/java/org/elasticsearch/common/util/CollectionUtils.java:
```diff
@@ -145,6 +150,18 @@ public static boolean isEmpty(Object[] array) {
+     * @param coll  the collection to check, may be null
+     * @return true if empty or null
+     */
+    public static boolean isEmpty(final Collection<?> coll) {
```
I'd prefer not do have this and to do the null check at the call site to honest.

nik9000 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/Request.java:
```diff
@@ -583,6 +585,22 @@ static Request clusterPutSettings(ClusterUpdateSettingsRequest clusterUpdateSett
+        params.withTimeout(healthRequest.timeout());
+        params.withMasterTimeout(healthRequest.masterNodeTimeout());
+        params.withLocal(healthRequest.local());
+        params.putParam("level", "shards");
```
I see. Right now you request the whole thing because it is simpler that way. I think this is fine. I think if we decide we'd like to support `level` as something else then we can do it in a followup. What do you think, @javanna?

javanna replied:
I had missed this line, so my comment on this is a bit off, sorry about that :) this is actually perfectly backwards compatible with the transport client behaviour, as it gets exactly everything. I think that we could add the level parameter though so that we can properly filter what gets returned at REST like every client works.

Van0SS commented:
Thank you guys for the review!
@javanna 

> We could also think of moving the filtering logic to the transport action rather than the REST layer, given that we retrieve everything at the moment but only print out at the REST layer what was requested. But I would not do it now. We will just need to be carefully to provide the proper params to the response#toXContent when testing. Let me know if you have questions on that.

I agree that better do it on the transport level.

> We probably forgot, could you also add that parameter to the SPEC?

I've dug deeper and seems that master_time isn't used in Cluster Health at all . I didn't find usages of MasterNodeRequest::masterNodeTimeout() connected to TransportClusterHealthAction. Probably it should be in TransportClusterHealthAction::masterOperation - request.timeout() for request.local() and request.masterNodeTimeout() for !request.local().
Maybe I'm wrong, please check it.

I also have a question about integration tests - do we run tests only on 1 node cluster, so I can check that numberOfNodes == 1 or it not always the case?

Van0SS commented:
@javanna @nik9000 
Also found out that I forget to add wait_for_event (i.e. priority) which also not in the docs. Do we need it? If yes, how better document it? Seems that if wait_for_event  not set then priority for the task even more than IMMEDIATE because it runs in the same Thread without scheduling.

javanna commented:
@Van0SS `masterNodeTimeout` is used in the base class `TransportMasterAction`, in the code that ends up calling `TransportClusterHealthAction#masterOperation`.

> do we run tests only on 1 node cluster, so I can check that numberOfNodes == 1 or it not always the case

I think so, yes. Do try running tests though and see what happens ;)

> Also found out that I forget to add wait_for_event (i.e. priority) which also not in the docs.

True, it's probably undocumented for a reason, the important bit is that it's in the SPEC and properly parsed in the REST action. We can see whether it needs to be documented in the Elasticsearch docs later, as a followup.

Thank you!!

Van0SS commented:
@javanna Fixes are here!
Also, added more comprehensive tests (seems that 'gradle check' is okay with one node asserts)
Added level parameter
Added documentation

javanna commented on test/framework/src/main/java/org/elasticsearch/test/AbstractXContentTestCase.java:
```diff
@@ -65,7 +62,8 @@
             assertEqualsConsumer.accept(testInstance, parsed);
             if (assertToXContentEquivalence) {
-                assertToXContentEquivalent(shuffled, XContentHelper.toXContent(parsed, xContentType, false), xContentType);
+                assertToXContentEquivalent(shuffled, XContentHelper.toXContent(parsed, xContentType, toXContentParams,false),
```
missing space in `toXContentParams,false` between the two arguments, here and anywhere else.

Van0SS commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/ClusterClientDocumentationIT.java:
```diff
@@ -156,7 +164,7 @@ public void testClusterUpdateSettingsAsync() throws Exception {
 
             // tag::put-settings-execute-listener
-            ActionListener<ClusterUpdateSettingsResponse> listener = 
+            ActionListener<ClusterUpdateSettingsResponse> listener =
```
Sorry, my IDE did it.

Van0SS commented on server/src/main/java/org/elasticsearch/cluster/health/ClusterIndexHealth.java:
```diff
@@ -173,19 +258,9 @@ public void writeTo(final StreamOutput out) throws IOException {
-
     @Override
     public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+        builder.field(INDEX, getIndex());
```
Right, I added in PR description:

For discussion in this PR:

- Serialize in XContent object name (in this case 'index' and 'shardId')
  - Pros: Objects can be XContented without level-up name and still have all fields; Easier to parse (including for any other clients)
  - Cons: Non-breaking but API change; A little overhead

So main Pros: each object can be tested for serialization without parent object and can be reused in other classes. But parsing in general will definitely work with a name in Context.

javanna replied:
ok I had missed that bit in the description. I would not make this change.  I don't completely follow how this makes things easy, it is just one less field that is printed out, but this is a fragment that is usually contained in its own object. Maybe we could rather move the object creation part to this toXContent method then. I would assume that this object is only ever printed out as part of its index object parent, so that might work.

Van0SS replied:
So, if I add `startObject()` and `endObject()` to this `toXContent` and change interface to `ToXContentObject`, can I leave `builder.field(INDEX, getIndex())` here? Without this field default XContent serialization/parsing tests will not work.

javanna replied:
no, the idea would be to add the `startObject(INDEX)` that is now done externally, as well as the corresponding endObject, but that doesn't make this class  a ToXContentObject, it is still a fragment. I would not add the additional index field. Let me know if I am missing anything.

Van0SS replied:
I finally got it, thanks. I didn't find a good example of using Fragment which writes `startObject(name)` by itself, has parsing and also has parsing tests. If you can point out, it would be great. So, I wrote this:
https://gist.github.com/Van0SS/9e6e8696a86b5300e55453bd86c9105d
What do you think?

javanna replied:
It is true that we tend to test mainly requests and responses which print out complete objects. I don't think fragments differ that much from objects though, given that they both implement `ToXContent`. The base test class requires a `ToXContent` and when calling `toXContent` it will wrap the output in an anonymous object when needed (see `XContentHelper#toXContent`). What issues did you encounter? Could you push what you have so I can review?

Van0SS commented on server/src/test/java/org/elasticsearch/cluster/health/ClusterShardHealthTests.java:
```diff
@@ -0,0 +1,46 @@
+        return new ClusterShardHealth(id, randomFrom(ClusterHealthStatus.values()),  randomInt(1000), randomInt(1000),
+            randomInt(1000), randomInt(1000), randomBoolean());
+    }
+
```
`AbstractStreamableXContentTestCase` has `supportsUnknownFields() { return true; }`
But `AbstractSerializingTestCase` has `supportsUnknownFields() { return false; }`
I think they should be the same, but not sure which default behavior is better, I would prefer `true`.

Van0SS replied:
@javanna I thought I added everywhere when it's not `true` by default. But I forgot to add it here, sorry. I also try to follow: "override only when needed to change default to true".

Van0SS commented on server/src/test/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponsesTests.java:
```diff
@@ -80,13 +89,60 @@ private void assertClusterHealth(ClusterHealthResponse clusterHealth) {
+    private static final Pattern SHARDS_IN_XCONTENT = Pattern.compile("^indices\\." + "\\w+" + "\\.shards$");
+
+    protected Predicate<String> getRandomFieldsExcludeFilter() {
+        return field -> "indices".equals(field) || SHARDS_IN_XCONTENT.matcher(field).find();
```
It can be done as manual checking if String is "indices.RANDOMINDEXNAME.shards" then return true, but I thinks regex is better. I will add explanation.

Van0SS commented on test/framework/src/main/java/org/elasticsearch/test/AbstractXContentTestCase.java:
```diff
@@ -42,18 +42,15 @@
+            boolean supportsUnknownFields, String[] shuffleFieldsExceptions, Predicate<String> randomFieldsExcludeFilter,
+            CheckedBiFunction<XContent, BytesReference, XContentParser, IOException> createParserFunction,
+            CheckedFunction<XContentParser, T, IOException> parseFunction, BiConsumer<T, T> assertEqualsConsumer,
+            boolean assertToXContentEquivalence, ToXContent.Params toXContentParams) throws IOException {
```
Probably... But is there a standard for method parameters indentation?

javanna replied:
not at the moment, we just try not to make unnecessary changes that may clutter diffs. We will introduce automatic formatting at some point, not there yet though.

javanna commented on test/framework/src/main/java/org/elasticsearch/test/AbstractXContentTestCase.java:
```diff
@@ -127,4 +125,8 @@ protected boolean assertToXContentEquivalence() {
         return Strings.EMPTY_ARRAY;
     }
+
+    protected ToXContent.Params getToXContentParams() {
```
no worries, small thing.

javanna replied:
and this one too?

Van0SS commented on server/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthRequest.java:
```diff
@@ -45,6 +45,12 @@
+     * Can be one of 'cluster', 'indices' or 'shards'. Controls the details level of the health information returned.
+     * The default value is 'shards' so it backwards compatible with the transport client behaviour.
+     */
+    private String level = "shards";
```
I would prefer followup solution.

Van0SS commented on server/src/test/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponsesTests.java:
```diff
@@ -80,13 +89,60 @@ private void assertClusterHealth(ClusterHealthResponse clusterHealth) {
+            randomDoubleBetween(0d, 100d, true), randomFrom(ClusterHealthStatus.values()), indices);
+
+        return new ClusterHealthResponse(randomAlphaOfLengthBetween(1, 10), randomInt(100), randomInt(100), randomInt(100),
+            TimeValue.timeValueMillis(randomInt(10000)), randomBoolean(), stateHealth);
```
`ESTestCase::randomTimeValue` returns `String`. I don't think that:
`TimeValue.parseTimeValue(randomTimeValue(), "property")` is better.

javanna replied:
you are right it is not.

Van0SS commented:
@javanna Round 3 changes were committed:
- Fix documentation
- Move health calculation logic from constructors to HealthAction
- Index, Shards: Remove toXContent name serialization.
startObject(name) inside the fragment
- Add missing overrides
- Use 8 spaces continuation indent  #29440

javanna commented on test/framework/src/main/java/org/elasticsearch/test/AbstractStreamableXContentTestCase.java:
```diff
@@ -64,4 +64,8 @@ protected boolean supportsUnknownFields() {
         return Strings.EMPTY_ARRAY;
     }
+
+    protected ToXContent.Params getToXContentParams() {
```
this is still missing?

javanna commented:
heya @Van0SS I haven't heard back after my last review, would you have a chance to address the comments? Otherwise please let me know what I can do to move this forward. Thanks!

Van0SS commented:
@javanna Hey, sorry, being busy, hope will update PR soon.

javanna commented:
no worries @Van0SS thank you! let me know if can help.

javanna commented:
test this please

javanna commented:
retest this please

javanna commented:
retest this please

javanna commented:
add to whitelist

javanna commented:
retest this please

nik9000 commented on docs/java-rest/high-level/cluster/health.asciidoc:
```diff
@@ -0,0 +1,189 @@
+--------------------------------------------------
+include-tagged::{doc-tests}/ClusterClientDocumentationIT.java[health-request-local]
+--------------------------------------------------
+<1> Non-master node can be used for this request. Defaults to `false`
```
Do you want to do `wait_for_events` here?

nik9000 commented on server/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java:
```diff
@@ -55,16 +151,28 @@ public ClusterHealthResponse(String clusterName, String[] concreteIndices, Clust
 
     public ClusterHealthResponse(String clusterName, String[] concreteIndices, ClusterState clusterState, int numberOfPendingTasks,
-                                 int numberOfInFlightFetch, int delayedUnassignedShards, TimeValue taskMaxWaitingTime) {
+        int numberOfInFlightFetch, int delayedUnassignedShards, TimeValue taskMaxWaitingTime) {
```
Can you add the indent back? It is much easier to read if the parameters don't line up with the code.

nik9000 commented on server/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java:
```diff
@@ -55,16 +151,28 @@ public ClusterHealthResponse(String clusterName, String[] concreteIndices, Clust
+     * For XContent Parser and serialization tests
+     */
+    ClusterHealthResponse(String clusterName, int numberOfPendingTasks, int numberOfInFlightFetch, int delayedUnassignedShards,
+        TimeValue taskMaxWaitingTime, boolean timedOut, ClusterStateHealth clusterStateHealth) {
```
Same here.

javanna commented:
thanks a lot @Van0SS !

Van0SS commented:
Thank you @javanna for finishing it!
